File: csharp/ToolGood.Words/internals/TrieNode.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace ToolGood.Words.internals
{
    public sealed class TrieNode //:IDisposable
    {
        public int Index;
        public int Layer;
        public bool End { get { return Results != null; } }
        public char Char;
        public List<int> Results;
        public Dictionary<char, TrieNode> m_values;
        public TrieNode Failure;
        public TrieNode Parent;
        public bool IsWildcard;
        public int WildcardLayer;
        public bool HasWildcard;


        public TrieNode Add(char c)
        {
            TrieNode node;
            if (m_values == null) {
                m_values = new Dictionary<char, TrieNode>();
            } else if (m_values.TryGetValue(c, out node)) {
                return node;
            }
            node = new TrieNode();
            node.Parent = this;
            node.Char = c;
            m_values[c] = node;
            return node;
        }

        public void SetResults(int index)
        {

            if (Results == null) {
                Results = new List<int>();
            }
            Results.Add(index);
        }
        /// <summary>
        /// 伪释放
        /// </summary>
        public void Dispose()
        {
            if (Results!=null) {
                Results.Clear();
                Results = null;
            }
            if (m_values!=null) {
                m_values.Clear();
                m_values = null;
            }
            Failure =null;
            Parent = null;
        }
    }
}

File: java/toolgood.words/src/main/java/toolgood/words/internals/TrieNode.java
package toolgood.words.internals;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public final class TrieNode implements Comparable<TrieNode> {

    public int Index;
    public int Layer;
    public char Char;
    public List<Integer> Results;
    public HashMap<Character, TrieNode> m_values;
    public TrieNode Failure;
    public TrieNode Parent;
    public boolean IsWildcard;
    public int WildcardLayer;
    public boolean HasWildcard;


    public TrieNode() {
        // m_values = new HashMap<Character, TrieNode>();
        // Results = new ArrayList<Integer>();
    }
    public boolean End(){
        return Results!=null;
    }

    public TrieNode Add(final Character c) {
        if(m_values==null){
            m_values = new HashMap<Character, TrieNode>();
        }
        if (m_values.containsKey(c)) {
            return m_values.get(c);
        }
        final TrieNode node = new TrieNode();
        node.Parent = this;
        node.Char = c;
        m_values.put(c, node);
        return node;
    }

    public void SetResults(final Integer index) {
        if(Results==null){
            Results = new ArrayList<Integer>();
        }
        if (Results.contains(index) == false) {
            Results.add(index);
        }
    }

    @Override
    public int compareTo(final TrieNode o) {
        return this.Layer - o.Layer  ;
    }
    /**
     * 伪释放
     */
    public void Dispose()
    {
        if (Results!=null) {
            Results.clear();
            Results = null;
        }
        if (m_values!=null) {
            m_values.clear();
            m_values = null;
        }
        Failure =null;
        Parent = null;
    }
}
